### PR TITLE
Backport firmware & OS Upgrade to bionic

### PR DESCRIPTION
**Pull request by mmstick**
Reference #1

- [x] Add the `Computer` and `Model` labels to the info panel.
- [x] Lock the firmware upgrade button behind a GtkLockButton / GPermission.

## Firmware

- [x] Add the firmware upgrade box to the info panel
- [x] Fetch BIOS model and version from system76-firmware daemon
- [x] Download digest and changelog from system76-firmware daemon
- [x] Initialize Firmware updates dialog
- [x] Integrate changelog information into the dialog
- [x] Schedule firmware updates with system76-firmware daemon

## OS Upgrade

- [x] Add pop-upgrade daemon support
- [x] Add an OS upgrade button for asking the daemon to initiate an upgrade
  - [x] Attempt any required system repairs in advance, including mounting the recovery partition
  - [x] Perform that upgrade with the recovery partition, if available
  - [x] Upgrade the recovery partition in advance.
  - [x] Ask to upgrade the recovery partition if required, but only when required.
- [x] Only display the upgrade as an option if an upgrade is available
  - [x] Display the version being upgraded to in the label, if available
- [x] Enable the lock button to unlock the upgrade option as well
- [x] Listen for upgrade signals and update the progress bar & label as progress signals are received
- [x] Handle the scenario where the upgrade daemon's status changed before the upgrade daemon received the final result.

## OS Refresh

- [x] Add pop-upgrade daemon support

Screenshot from 2019-05-06 12-04-06

Screenshot from 2019-05-06 12-06-32

Screenshot from 2019-05-06 09-59-39

Screenshot from 2019-05-07 14-49-09

### CODE REVIEW

**Review by brs17 — CHANGES_REQUESTED** [review]
Clicking away from the `About` tab in gnome-control-center crashes(Seg faults) the application. 

Log message when selecting the `Details` tab:
```
(gnome-control-center:3354): info-cc-panel-WARNING **: 18:05:46.935: failed to call ReleaseCheck on PopUpgrade: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name com.system76.PopUpgrade was not provided by any .service files
```
Then, `Segmentation fault (core dumped)`

**Comment by brs17** [comment]
Highlighted underneath "OS Upgrade and Refresh": "No upgrades available" there seems to be a second highlightable widget that in the following screenshot I have highlighted. 
Screenshot from 2019-04-04 18-09-40

EDIT: This second widget is probably the space for the `Refresh` option. This probably is fixed with the rebasing with the refresh install feature that was just merged.

**Comment by brs17** [comment]
When no upgrades are available, I have no idea what the `Unlock` /`Lock` is supposed to do.

EDIT: Well I see in the authentication menu that it is for firmware and OS upgrades but I see no changes on the page...I'm not sure, but it seems odd to have a button that essentially does nothing in a specific use case (the use case that is more often the case than not)

**Comment by brs17** [comment]
If this is supposed to replace the system76-firmware application, shouldn't we list the installed version of the BIOS/EC/EC2/ME regardless of an update being available? The system76-firmware application also shows the changelog...I'm not sure if this should be listed on the About page (in its current layout), but should probably be provided somewhere.

File: debian/patches/s76_firmware.patch [hunk 1]
```diff
@@ -0,0 +1,1099 @@
...
++  GtkButton           *firmware_button,
++  GtkLabel            *firmware_label,
++  gchar              **firmware_digest,
++  gchar              **firmware_changelog
++) {
++  gtk_widget_hide (GTK_WIDGET (firmware_button));
++  const gchar *label;
++
++  if (g_file_test ("/sys/firmware/efi", G_FILE_TEST_IS_DIR)) {
++    label = _("No Updates Available");
```
Comment (brs17) [inline debian/patches/s76_firmware.patch]: `U` and `A` should be lowercase:
`No updates available`

**Comment by mmstick** [comment]
We could list them, if that's desired. The pop-upgrade daemon does have to be installed for the os upgrade feature to work. Soon there will be a "Refresh OS" button in there to give that unlock button more of a purpose. Segfault is likely due to a bad instance of `free ()`.

**Comment by brs17** [comment]
@mmstick My thought is that the changelog provides the user an understanding for what an update contains. I would think this will help convince users of the importance and the need to run up to date firmware.

**Comment by WatchMkr** [comment]
The change log description of the firmware update is on the confirmation page. Please review the original design for copy as well.

**Comment by WatchMkr** [comment]
@mmstick what does the unlock switch do for the "Refresh OS" option?

**Comment by mmstick** [comment]
None of the buttons in the panel can be clicked until the panel is unlocked. Same behavior as in the "Users" panel, which requires the panel to be unlocked to make changes to user accounts.

**Comment by brs17** [comment]
The buttons look cut off on the bottom of the button:
Screenshot from 2019-04-08 18-33-57

I noticed that the object containing the text of what is happening changes size based on the content within:
Screenshot from 2019-04-08 18-45-44
Screenshot from 2019-04-08 18-45-54

After the updates were downloaded, I had the "About" page which said that a reboot was required. I tried clicking on that label, and gnome-control-center crashed. 

I then rebooted and was brought into a 18.04 recovery partition. Getting back to the user partition, I was prompted in gnome-control-center that an upgrade was available.

**Comment by brs17** [comment]
- I saw gnome-control-center crash after finishing with "Upgrade from 18.04->18.10".

- The recovery partition update works, although the About page still says an update is available.

**Comment by mmstick** [comment]
There are some segfaults at the end. Once you're in the recovery partition, it is possible to install the installer PR to try out the upgrade from the recovery partition. The upgrade button in GCC is doing a recovery upgrade when it detects that a recovery partition exists, even if the recovery partition is not mounted at the time of the upgrade attempt.

**Comment by WatchMkr** [comment]
We won’t want verbose messages here. Please see copy for the feature here:

https://github.com/pop-os/gnome-control-center/issues/8

I believe it covers all copy scenarios for the feature (at least in the Gnome UI).

**Comment by mmstick** [comment]
I'll remove handling of the specific package events for GCC. These events are mainly for the CLI frontend.

**Comment by brs17** [comment]
- With the latest changes, I notice that it is taking significantly longer to select/load the "Details" page than without the patch.

- While a mockup has not been provided of how to display the firmware versions, I would think we should not provide the firmware versions in textboxes as I see in the following screenshot, but rather labels(I'm assuming) like the information provided higher up on the page. I think we should also strive to keep similar formatting. I notice that `bios`, `ec, `ec2`, `me` are all bolded while titles like `Computer`, `Model`, `Memory`, are not. Also capitalization should be consistent.

firmware

- I also noticed that the version for the M.E. from the screenshot (11.8.50.3425) is not what is shown in the firmware daemon nor the BIOS (disabled/0.0.0.0)

**Comment by brs17** [comment]
Selecting "Reflash" caused a segfault in gnome-control-center.

Appropriate journald logs:

[logs.log](https://github.com/pop-os/gnome-control-center/files/3138724/logs.log)

**Comment by brs17** [comment]
I reverted the firmware to an older version so that I could test upgrading firmware. I did not see firmware versions, nor the ability to upgrade.

Screenshot from 2019-05-02 10-31-37

I then rebooted and got the information I was expecting.

**Comment by brs17** [comment]
The firmware update popup is missing the information regarding the new firmware available.

I am also thinking the `Reboot and Install` button should probably be an affirmative color as we want users to upgrade to the latest firmware, but that is definitely not my area of "expertise" perhaps @isantop or @WatchMkr have a thought about it.

Screenshot from 2019-05-02 10-58-44

**Comment by brs17** [comment]
Upon pressing the `Reboot and Install` button, gnome-control-center closed. 

Log line from journald from time of crash

```
May 02 11:09:59 pop-os kernel: traps: gnome-control-c[2106] general protection ip:7fa8fed476a6 sp:7fffd561c020 error:0 in libgio-2.0.so.0.5600.4[7fa8fec7a000+195000]
```

**Comment by brs17** [comment]
On a BIOS system, gnome-control-center segfaults and I am unable to open the application.

**Comment by mmstick** [comment]
The crashes should be fixed now

**Comment by mmstick** [comment]
Updated screenshots. Almost finished with this PR

**Comment by brs17** [comment]
EDIT: I think the `Refresh OS` button should be red instead of green as it is a destructive action.

**Comment by brs17** [comment]
I notice that the M.E. version is listed but the M.E. is disabled so it should say `disabled`. In the firmware updater tool the m.e. is listed as disabled.

**Comment by brs17** [comment]
When selecting the `Recovery update available` this should update the recovery partition to the latest iso/image of the current OS installed.

When selecting `Upgrade from 1804 to 19.04`, I am first prompted to upgrade the recovery partition. Since this upgrade would turn the recovery partition into a 19.04 image, I am not sure we want to prompt the user as this is a required part of upgrading the system.

When the download/upgrade process is being shown in GCC, the `Refresh OS` option says `Release upgrade in progress - please wait`. If possible I would say that this option should be hidden during the installation process and only shown after the installation is complete.

I also noticed an odd issue with the inability to resize the About page. @mmstick seemed to be aware of what the fix should be.

**Comment by brs17** [comment]
UPDATE: After a discussion with the entire team, the firmware information is going to be re-implemented in a library that can then be added to gnome-control-center or used in the Ubuntu offering.

**Comment by mmstick** [comment]
Finished the firmware widget library implementation. Now working on a library implementation of the upgrade portion. Closing this in the meantime, as the two libraries will be in two different pull requests.

**Pull request closed without merging**